src/wallet/feebumper.cpp

```diff
@@ -0,0 +1,275 @@
+    :
+    txid(txidIn),
+    nOldFee(0),
+    nNewFee(0)
+{
+    vErrors.clear();
+    bumpedTxid.SetNull();
+    AssertLockHeld(pWallet->cs_wallet);
+    if (!pWallet->mapWallet.count(txid)) {
+        vErrors.push_back("Invalid or non-wallet transaction id");
+        currentResult = BumpFeeResult::INVALID_ADDRESS_OR_KEY;
+        return;
+    }
+    auto it = pWallet->mapWallet.find(txid);
+    const CWalletTx& wtx = it->second;
+
+    if (pWallet->HasWalletSpend(txid)) {
+        vErrors.push_back("Transaction has descendants in the wallet");
+        currentResult = BumpFeeResult::MISC_ERROR;
+        return;
+    }
+
+    {
+        LOCK(mempool.cs);
+        auto it_mp = mempool.mapTx.find(txid);
+        if (it_mp != mempool.mapTx.end() && it_mp->GetCountWithDescendants() > 1) {
+            vErrors.push_back("Transaction has descendants in the mempool");
+            currentResult = BumpFeeResult::MISC_ERROR;
+            return;
+        }
+    }
+
+    if (wtx.GetDepthInMainChain() != 0) {
+        vErrors.push_back("Transaction has been mined, or is conflicted with a mined transaction");
+        currentResult = BumpFeeResult::INVALID_ADDRESS_OR_KEY;
+        return;
+    }
+
+    if (!SignalsOptInRBF(wtx)) {
+        vErrors.push_back("Transaction is not BIP 125 replaceable");
+        currentResult = BumpFeeResult::INVALID_ADDRESS_OR_KEY;
+        return;
+    }
+
+    if (wtx.mapValue.count("replaced_by_txid")) {
+        vErrors.push_back(strprintf("Cannot bump transaction %s which was already bumped by transaction %s", txid.ToString(), wtx.mapValue.at("replaced_by_txid")));
+        currentResult = BumpFeeResult::INVALID_REQUEST;
+        return;
+    }
+
+    // check that original tx consists entirely of our inputs
+    // if not, we can't bump the fee, because the wallet has no way of knowing the value of the other inputs (thus the fee)
+    if (!pWallet->IsAllFromMe(wtx, ISMINE_SPENDABLE)) {
+        vErrors.push_back("Transaction contains inputs that don't belong to this wallet");
+        currentResult = BumpFeeResult::INVALID_ADDRESS_OR_KEY;
+        return;
+    }
+
+    // figure out which output was change
+    // if there was no change output or multiple change outputs, fail
+    int nOutput = -1;
+    for (size_t i = 0; i < wtx.tx->vout.size(); ++i) {
+        if (pWallet->IsChange(wtx.tx->vout[i])) {
+            if (nOutput != -1) {
+                vErrors.push_back("Transaction has multiple change outputs");
+                currentResult = BumpFeeResult::MISC_ERROR;
+                return;
+            }
+            nOutput = i;
+        }
+    }
+    if (nOutput == -1) {
+        vErrors.push_back("Transaction does not have a change output");
+        currentResult = BumpFeeResult::MISC_ERROR;
+        return;
+    }
+
+    // Calculate the expected size of the new transaction.
+    int64_t txSize = GetVirtualTransactionSize(*(wtx.tx));
+    const int64_t maxNewTxSize = CalculateMaximumSignedTxSize(*wtx.tx);
```

Comment:
There is a minor bug here. This needs to pass `pWallet` into `CalculateMaximumSignedTxSize` so it can stop using `pwalletMain`. When making this change, I also had to make the `CWallet::DummySignTx` method const. Feel free to use my complete fix: https://github.com/ryanofsky/bitcoin/commit/731ca5c652d42806e82892695f4e43082c9d6a20